radis/misc/profiler.py

```diff
@@ -0,0 +1,126 @@
+"""
+`Profiler` class, for printing time spent on various steps during Spectrum
+calculation under :py:class:`~radis.lbl.factory.SpectrumFactory` based on verbose level.
+
+Also stores Spectrum calculation time dependent parameters, under the attribute
+    :py:attr:`~radis.misc.profiler.Profiler.dict_time`
+
+Routine Listing
+---------------
+
+- :meth:`~radis.misc.profiler.Profiler.start`
+- :meth:`~radis.misc.profiler.Profiler.stop`
+- :meth:`~radis.misc.profiler.Profiler._print`
+
+-------------------------------------------------------------------------------
+"""
+
+from time import time
+
+from radis.misc.printer import printg
+
+
+class Profiler(object):
+    """A class to store Spectrum calculation time dependent parameters, under the attribute
+    :py:attr:`~radis.misc.profiler.Profiler.dict_time` of
+    :py:class:`~radis.lbl.factory.SpectrumFactory`.
+
+    It also hold functions to print all the entities based on verbose value.
+
+    See Also
+    --------
+
+    :py:attr:`~radis.lbl.loader.DatabankLoader.input`,
+    :py:attr:`~radis.misc.profiler.Profiler,
+
+    """
+
+    def __init__(self, verbose):
+        super(Profiler, self).__init__()
+
+        # Dev: Init here to be found by autocomplete
+        self.initial = {}
+        self.dict_time = {}
+        self.verbose = verbose
+        self.relative_time_percentage = {}
+
+    def start(self, key, verbose):
+        if __debug__:
+            self.initial[key] = {
+                "start_time": time(),
+                "verbose": verbose,
+            }
+
+    def stop(self, key, details=""):
+        if __debug__:
+            items = self.initial.pop(key)
+            # Storing time with verbose level
+            self.dict_time[key] = [time() - items["start_time"], items["verbose"]]
+            if self.verbose >= items["verbose"]:
+                self._print(items["verbose"], details, key)
+
+    def _print(self, verbose, details, key=None):
+        if verbose == 1:
+            if key == None:
+                print(details)
+            else:
+                print("{0:.2f}s -".format(self.dict_time[key][0]), details)
+        elif verbose >= 2:
+            if key == None:
+                printg("..." * (verbose - 1), details)
+            else:
+                printg(
+                    "..." * (verbose - 1),
+                    "{0:.2f}s -".format(self.dict_time[key][0]),
+                    details,
+                )
+
+    def percentage_distribution(self):
```

Comment:
@dcmvdbekerom added percentage time distribution for steps involved in each verbose level. I'm storing it in a Spectrum object and the result will be as follow :

`spectrum.get_conditions()['profiler']` :

```
{1: [('spectrum_calculation', [100.0, 1.119952917098999])],
 2: [('spectrum_calc_before_obj', [56.97151990399073, 1.061187982559204]),
  ('calc_noneq_population', [19.74545153399409, 0.36779141426086426]),
  ('check_non_eq_param', [9.267285489209133, 0.17261838912963867]),
  ('fetch_energy_6', [7.526554291483005, 0.14019441604614258]),
  ('generate_spectrum_obj', [3.1442882049576717, 0.05856752395629883]),
  ('calc_line_broadening', [1.9036095507002235, 0.03545784950256348]),
  ('calc_weight_trans', [0.9310497852245717, 0.017342329025268555]),
  ('non_eq_linestrength', [0.14664811709423994, 0.0027315616607666016]),
  ('calc_other_spectral_quan', [0.11783560844632739, 0.0021948814392089844]),
  ('calc_emission_integral', [0.1084405034496291, 0.0020198822021484375]),
  ('calc_hwhm', [0.10745491341591554, 0.002001523971557617]),
  ('calc_lineshift', [0.029862098034464674, 0.0005562305450439453])],
 3: [('part_function', [93.1809938338774, 0.3619859218597412]),
  ('DLM_convolve', [4.159662498841589, 0.0161592960357666]),
  ('population', [1.4629428122355217, 0.005683183670043945]),
  ('corrected_population_se', [0.6323236898293653, 0.0024564266204833984]),
  ('precompute_DLM_lineshapes', [0.37977472509600235, 0.0014753341674804688]),
  ('DLM_closest_matching_line', [0.08273049926137867, 0.00032138824462890625]),
  ('DLM_Distribute_lines', [0.050571165720605354, 0.0001964569091796875]),
  ('map_part_func', [0.04965057411161376, 0.00019288063049316406]),
  ('DLM_Initialized_vectors', [0.0013502010265210168, 5.245208740234375e-06])]}
```
So its ```
{        {
            verbose_level: [
                ("key_1", [percentage_time, time]),
                ("key_2", [percentage_time, time]),
            ]
        }
```
and in sorted manner!
Your feedback!

Reply:
Can you **not** store the % ? Less data on the Spectrum object eventually will be more readable / easier to serialize. 

However, the function that Prints() should show the percentages (as you've done). 
Eventually you may want a print() functino in SpectrumFactory or Spectrum ; that parses the dictionary of computed times and renders it properly. 

(Basically : storing times; and printing them; should have a different philosophy : store only the minimum ; and print which as much details informatino as the user wants )

Reply:
Actually, I'm not storing the percentage data in the Spectrum object. If you see I'm updating only the method and its time using:
```
   conditions.update(
            {
                "profiler": {
                    i: self.profiler.dict_time[i][0] for i in self.profiler.dict_time
                }
            }
        )
```
The whole calculation thing is right now stored in a different variable `sf.profiler.relative_time_percentage`, so anyone who wants the full detail can directly access it. 
I'm a little confused by the print function you have mentioned. We use the `get_conditions()` method to get the parameters, so a similar function `self.print_details` can be used to print all the details?

Reply:
Yes for the print, but we can always add it later depending on your needs. For now I think thr implementation is good and ready! Merging if you think it's fine

Reply:
Pushing one last commit!

Reply:
@erwanp I was thinking of adding a column 'name' or 'id' (eg - 'CO_9000_10000') which will be same for a single Spectrum Factory, so that it will be easier for me to distinguish between various SF in the csv generated file. The file name gets incremented when computing several spectrum from Spectrum Factory so there isn't any unique field to distinguish 1 SF to another.

Reply:
you can also give a name to the Spectra, even at runtime (eq_spectrum(name=...). Maybe that's as convenient ; and it's already implemented.